respy/exogenous_processes.py

```diff
@@ -0,0 +1,139 @@
+"""Utils for the inclusion of exogenous processes."""
+import functools
+
+import numpy as np
+import pandas as pd
+from scipy import special
+
+from respy.parallelization import parallelize_across_dense_dimensions
+from respy.shared import get_exogenous_from_dense_covariates
+from respy.shared import load_objects
+from respy.shared import pandas_dot
+
+
+def compute_transition_probabilities(
+    states, transit_keys, optim_paras, dense_key_to_dense_covariates
+):
+    """Insert Docstring."""
+    exogenous_processes = optim_paras["exogenous_processes"]
+
+    dense_key_to_exogenous = {
+        key: get_exogenous_from_dense_covariates(
+            dense_key_to_dense_covariates[key], optim_paras
+        )
+        for key in transit_keys
+    }
+
+    # Compute the probabilities for every exogenous process.
+    probabilities = []
+    for exog_proc in exogenous_processes:
+
+        # Create the dot product of covariates and parameters.
+        x_betas = []
+        for params in exogenous_processes[exog_proc].values():
+            x_beta = pandas_dot(states[params.index], params)
+            x_betas.append(x_beta)
+
+        probs = special.softmax(np.column_stack(x_betas), axis=1)
+        probabilities.append(probs)
+
+    # Prepare full Dataframe. If issues arrise we might want to switch typed dicts
+    df = pd.DataFrame(index=states.index)
+
+    for dense in dense_key_to_exogenous:
+        array = functools.reduce(
+            np.multiply,
+            [
+                probabilities[proc][:, val]
+                for proc, val in enumerate(dense_key_to_exogenous[dense])
+            ],
+        )
+        df[str(dense)] = array
+
+    return df
+
+
+@parallelize_across_dense_dimensions
+def weight_continuation_values(
+    complex_, options, continuation_values, transit_key_to_choice_set
+):
+    """Insert Docstring."""
+    transition_df = load_objects("transition", complex_, options)
+
+    choice_set = complex_[1]
+    # Reconsider this setup
+    choice_positons = {
+        key: _get_representation_cols(value, choice_set)
+        for key, value in transit_key_to_choice_set.items()
+    }
+    continuation_values_adjusted = {
+        ftr_key: continuation_values[int(ftr_key)][
```

Comment:
Very minor, but what does `ftr_key` stand for here? Would it be possible to write out the word?

Reply:
future_key since these keys represent dense states one could be in the next period. I used it because we already use quite long names. But you are right. Ill change it to future_key